xtractmime/main.py

```diff
@@ -0,0 +1,90 @@
+from typing import List, Union
+
+_APACHE_TYPES_STR = [
+    "text/plain",
+    "text/plain; charset=ISO-8859-1",
+    "text/plain; charset=iso-8859-1",
+    "text/plain; charset=UTF-8",
+]
+
+_APACHE_TYPES_BYTES = [
+    "746578742f706c61696e",
+    "746578742f706c61696e3b20636861727365743d49534f2d383835392d31",
+    "746578742f706c61696e3b20636861727365743d69736f2d383835392d31",
+    "746578742f706c61696e3b20636861727365743d5554462d38",
+]
+
+
+_WHITESPACE_BYTES = [b"\t", b"\r", b"\x0c", b"\n", b" "]
+
+
+# handling resource metadata
+def _should_check_for_apache_bug(supplied_type, http_origin):
+    if http_origin and (
+        supplied_type in _APACHE_TYPES_STR or supplied_type in _APACHE_TYPES_BYTES
+    ):
+        return True
+
+    return False
+
+
+# reading resource header
+def _read_resource_header(body):
+    buffer = b""
+    if len(body) < 1445:
+        buffer = body
+    else:
+        buffer = body[:1445]
+    return buffer
+
+
+# Matching a MIME type pattern
+def _is_match_mime_pattern(
+    input_bytes, byte_pattern, pattern_mask, lead_whitespace=None
+):
+    input_size, pattern_size, mask_size = (
+        len(input_bytes),
+        len(byte_pattern),
+        len(pattern_mask),
+    )
+
+    assert pattern_size == mask_size, "pattern's length should match mask's length"
+
+    if input_size < pattern_size:
+        return False
+
+    input_index, pattern_index = 0, 0
+
+    if lead_whitespace:
+        while (
+            input_index < input_size
+            and bytes([input_bytes[input_index]]) in _WHITESPACE_BYTES
```

Comment:
This looks overly complicated. You could turn `_WHITESPACE_BYTES` into `_WHITESPACE_BYTES = b"\t\r\x0c\n "`, and turn this into:

```suggestion
            and input_bytes[input_index] in _WHITESPACE_BYTES
```

Moreover, since the “contains” operation is the only one you need, turning `_WHITESPACE_BYTES` into a set would be best: `{b"\t", b"\r", b"\x0c", b"\n", b" "}` (or `set(b"\t\r\x0c\n ")`, for better readability).

I had read somewhere that lists could be faster for a short set of elements, but it turns out I was wrong, or at least sets seem to be much better in this specific scenario:

```python
>>> timeit.timeit(r'b" " in variable', r'variable = {b"\t", b"\r", b"\x0c", b"\n", b" "}', number=1_000_000)
0.027940084000874776
>>> timeit.timeit(r'b" " in variable', r'variable = b"\t\r\x0c\n "', number=1_000_000)
0.13945079399854876
>>> timeit.timeit(r'b"\t" in variable', r'variable = b"\t\r\x0c\n "', number=1_000_000)
0.1449901920004777
```

Reply:
`and input_bytes[input_index] in _WHITESPACE_BYTES` is failing as `input_bytes[input_index]` return the ascii value rather than return a byte value.